proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,237 @@
+    [plugins.cri.containerd.runtimes.slight-v1-0-0-beta1]
+      runtime_type = "io.containerd.slight-v1-0-0-beta1.v1"   # must use "-" rather than "."
+    [plugins.cri.containerd.runtimes.slight-v0-1-0-alpha1]
+      runtime_type = "io.containerd.slight-v0-1-0-alpha1.v1"   # must use "-" rather than "."
+    [plugins.cri.containerd.runtimes.slight-v0-2-0-alpha1]
+      runtime_type = "io.containerd.slight-v0-2-0-alpha1.v1"   # must use "-" rather than "."
+    ```
+3. The following shows example runtime classes that map handler names to the config runtime names from step 2. It also describes 3 types of nodeSelectors which can be used to provide differing levels of workload selectivity, specific version, major version, and presence of shim regardless of version.
+      ```yaml
+      # This runtime class maps "plugins.cri.containerd.runtimes.slight-v0-1-0-alpha1" to
+      # runtime_type = "io.containerd.slight-v0-1-0-alpha1.v1".
+      # 
+      # This node selector is intended to select only nodes that have the slight v0.1.0-alpha
+      # version of the slight shim. This is critical to enable a workload to target a specific
+      # shim version during nascent stages of shim development when forward and / or backward
+      # compatibility can not be guaranteed.
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-slight-v0-1-0-alpha1
+      handler: slight-v0-1-0-alpha1
+      scheduling:
+        nodeSelector:
+          wasmtime-slight-v0-1-0-alpha1: "true"
+      ---
+      # This runtime class maps "plugins.cri.containerd.runtimes.slight-v1-2-3" to
+      # runtime_type = "io.containerd.slight-v1-2-3.v1".
+      # 
+      # This node selector is intended to select any node that has a slight shim enabled
+      # within the v1 major version. As new nodes are introduced with newer versions in the
+      # v1 series (e.g. v1.3.4, v1.5.6, etc) a workload could specify a runtime class of 
+      # "wasmtime-slight-v1" and run on any node with a v1 series shim version installed.
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-slight-v1
+      handler: slight-v1-2-3
+      scheduling:
+        nodeSelector:
+          wasmtime-slight-v1: "true"
+      ---
+      # This runtime class maps "plugins.cri.containerd.runtimes.slight-v1" to
+      # runtime_type = "io.containerd.slight-v1-2-3.v1".
+      # 
+      # This node selector is intended select any node that has the slight shim enabled
+      # regardless of the version of the shim. As new versions are introduced, this
+      # node selector would allow for new versions of the shim to be introduced without
+      # having to update the pod workload to target a new shim version.
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-slight
+      handler: slight-v1
```

Comment:
Why does handler still have a major version while the name of the runtimeclass has no indication of what major version the shim is?

Reply:
Good catch. 
1. slight-v1 is not defined in the handlers.
2. The goal of this runtime class is to simply indicate the presence of the type of shim you'd like to run your workload, but does not give any assurance of the version, thus "wasmtime-slight". I tried to explain that in the comments above the yaml. I've added a note comment to the handler line.